services/src/main/java/com/google/daq/mqtt/mapping/MappingService.java

```
@@ -1,102 +1,126 @@
    */
-  public MappingService(String[] args) {
-    if (args.length != requiredArgumentLength) {
-      throw new IllegalArgumentException("Invalid arguments provided");
-    }
-
-    this.registryId = args[0];
-    this.projectId = args[1];
-    this.baseCloningDirectory = args[2];
-    this.projectSpec = args[3];
-    this.mappingFamily = args[4];
-    this.discoveryNodeDeviceId = args[5];
+  public MappingService(String projectTarget, String projectSpec, String siteModelBaseDir, String localOriginDir) {
+    super(SERVICE_NAME, SUBSCRIPTION_SUFFIX, projectTarget, siteModelBaseDir, localOriginDir);
+    this.projectSpec = projectSpec;
+    LOGGER.info("Starting Mapping Service for project {}, cloning to {}", projectTarget,
+        siteModelBaseDir);
   }
 
-  private void cloneRepo() {
-    if (!sourceRepository.clone(IMPORT_BRANCH)) {
-      throw new RuntimeException("Could not clone repository " + registryId);
-    }
-  }
+  @Override
+  protected void handleMessage(PubsubMessage message) {
+    try {
+      Map<String, Object> messageData = parseSourceRepoMessageData(message);
+      Integer eventNumber = (Integer) messageData.getOrDefault(EVENT_NUMBER_FIELD, 0);
+      LOGGER.info("Received event no. from message: {}", eventNumber);
+      if (eventNumber < 0) {
+        String mappingFamily = (String) messageData.getOrDefault(FAMILY_FIELD, "");
+        String registryId = message.getAttributesOrDefault(REGISTRY_ID_FIELD, "");
 
-  private void initialize() {
-    String repoCloneDir = Paths.get(baseCloningDirectory, registryId).toString();
-    sourceRepository = new SourceRepository(registryId, repoCloneDir, null, projectId, null);
+        if (registryId.isEmpty()) {
+          LOGGER.error("Registry Id not found for the message.");
+          return;
+        }
 
-  }
+        String discoveryNodeDeviceId = message.getAttributesOrDefault(DISCOVERY_NODE_DEVICE_ID_FIELD, "");
+        if (discoveryNodeDeviceId.isEmpty()) {
```

Do we want to process further if discovery node is empty? I think the Mapping Agent does some attribute checks later on the discovery node and further operations might not go through. Better to fail fast!